UefiPayloadPkg/PayloadLoaderPeim/FitPayloadLoaderPeim.c

```diff
@@ -128,6 +272,1018 @@ EFI_PEI_PPI_DESCRIPTOR  gPpiLoadFilePpiList = {
+  ASSERT (ParentNode > 0);
+
+  UPLParaNode = FdtAddSubnode (Fdt, ParentNode, "upl-params");
+  ASSERT (UPLParaNode > 0);
+
+  //
+  // Create CPU info FDT node
+  //
+  CpuHob = GetFirstHob (EFI_HOB_TYPE_CPU);
+  ASSERT (CpuHob != NULL);
+
+  Data16 = CpuToFdt16 (0);
+  Status = FdtSetProp (Fdt, UPLParaNode, "pci-enum-done", &Data16, sizeof (Data16));
+  ASSERT_EFI_ERROR (Status);
+
+  BootMode = GetBootModeHob ();
+
+  Data32 = CpuToFdt32 ((UINT32)CpuHob->SizeOfMemorySpace);
+  Status = FdtSetProp (Fdt, UPLParaNode, "addr-width", &Data32, sizeof (Data32));
+  ASSERT_EFI_ERROR (Status);
+
+  if (BootMode == BOOT_WITH_FULL_CONFIGURATION) {
+    Status = FdtSetProp (Fdt, UPLParaNode, "boot-mode", "normal", (UINT32)(AsciiStrLen ("normal")+1));
+  } else if (BootMode == BOOT_WITH_MINIMAL_CONFIGURATION) {
+    Status = FdtSetProp (Fdt, UPLParaNode, "boot-mode", "fast", (UINT32)(AsciiStrLen ("fast")+1));
+  } else if (BootMode == BOOT_WITH_FULL_CONFIGURATION_PLUS_DIAGNOSTICS) {
+    Status = FdtSetProp (Fdt, UPLParaNode, "boot-mode", "full", (UINT32)(AsciiStrLen ("full")+1));
+  } else if (BootMode == BOOT_WITH_DEFAULT_SETTINGS) {
+    Status = FdtSetProp (Fdt, UPLParaNode, "boot-mode", "default", (UINT32)(AsciiStrLen ("default")+1));
+  } else if (BootMode == BOOT_ON_S4_RESUME) {
+    Status = FdtSetProp (Fdt, UPLParaNode, "boot-mode", "s4", (UINT32)(AsciiStrLen ("s4")+1));
+  } else if (BootMode == BOOT_ON_S3_RESUME) {
+    Status = FdtSetProp (Fdt, UPLParaNode, "boot-mode", "s3", (UINT32)(AsciiStrLen ("s3")+1));
+  } else {
+    Status = FdtSetProp (Fdt, UPLParaNode, "boot-mode", "na", (UINT32)(AsciiStrLen ("na")+1));
+  }
+
+  ASSERT_EFI_ERROR (Status);
+
+  Status = FdtSetProp (Fdt, UPLParaNode, "compatible", "upl", (UINT32)(AsciiStrLen ("upl")+1));
+  ASSERT_EFI_ERROR (Status);
+
+  GuidHob = GetFirstGuidHob (&gUniversalPayloadBaseGuid);
+  if (GuidHob != NULL) {
+    PayloadBase = (UNIVERSAL_PAYLOAD_BASE *)GET_GUID_HOB_DATA (GuidHob);
+    Fit         = (VOID *)(UINTN)PayloadBase->Entry;
+    DEBUG ((DEBUG_INFO, "PayloadBase Entry = 0x%08x\n", PayloadBase->Entry));
+
+    Status       = AsciiSPrint (TempStr, sizeof (TempStr), "upl-images@%lX", (UINTN)(Fit));
+    UPLImageNode = FdtAddSubnode (Fdt, ParentNode, TempStr);
+
+    Data64 = CpuToFdt64 ((UINTN)Fit);
+    Status = FdtSetProp (FdtBase, UPLImageNode, "addr", &Data64, sizeof (Data64));
+  }
+
+  CustomNode = FdtAddSubnode (Fdt, ParentNode, "upl-custom");
+  ASSERT (CustomNode > 0);
+
+  HobPtr = GetHobList ();
+  Data64 = CpuToFdt64 ((UINT64)(EFI_PHYSICAL_ADDRESS)HobPtr);
+  Status = FdtSetProp (Fdt, CustomNode, "hoblistptr", &Data64, sizeof (Data64));
+  ASSERT_EFI_ERROR (Status);
+
+  return Status;
+}
+
+/**
+  It will build FDT for UPL consumed.
+  @param[in] FdtBase         Address of the Fdt data.
+  @retval EFI_SUCCESS        If it completed successfully.
+  @retval Others             If it failed to build required FDT.
+**/
+EFI_STATUS
+BuildFdtForUPL (

```

Comment:
Fdt table can be platform specific, so I would recommend that we move this function and relevant sub functions to a separate "BuildFdtLib" so different platforms can build their own Fdt table by providing their own library instance.

Reply:
resolved